posthog/temporal/batch_exports/spmc.py

```diff
@@ -243,34 +250,64 @@ async def start(
         writer = get_batch_export_writer(writer_format, self.flush, schema=schema, max_bytes=max_bytes, **kwargs)
 
         record_batches_count = 0
+        records_count = 0
+        record_batch_generator = self.generate_record_batches_from_queue(queue, producer_task)
 
         await self.logger.adebug("Starting record batch writing loop")
-        async with writer.open_temporary_file():
-            while True:
-                try:
-                    record_batch = queue.get_nowait()
-                    record_batches_count += 1
-                except asyncio.QueueEmpty:
-                    if producer_task.done():
-                        await self.logger.adebug(
-                            "Empty queue with no more events being produced, closing writer loop and flushing"
-                        )
-                        self.flush_start_event.set()
-                        # Exit context manager to trigger final flush
-                        break
-                    else:
-                        await asyncio.sleep(0)
-                        continue
-
-                record_batch = cast_record_batch_json_columns(record_batch, json_columns=json_columns)
-                await writer.write_record_batch(record_batch, flush=True)
-
-        for _ in range(record_batches_count):
-            queue.task_done()
-
-        await self.logger.adebug("Consumed %s records", writer.records_total)
+
+        writer._batch_export_file = writer.create_temporary_file()
+
+        while True:
+            try:
+                record_batch = await anext(record_batch_generator)
+            except StopAsyncIteration:
+                break
+
+            record_batches_count += 1
+            record_batch = cast_record_batch_json_columns(record_batch, json_columns=json_columns)
+
+            await writer.write_record_batch(record_batch, flush=False)
+
+            if writer.should_flush():
+                records_count += writer.records_since_last_flush
+
+                if multiple_files:
+                    await writer.close_temporary_file()
+                    writer._batch_export_file = writer.create_temporary_file()
+                else:
+                    await writer.flush()
+
+                for _ in range(record_batches_count):
+                    queue.task_done()
+                record_batches_count = 0
+
+        records_count += writer.records_since_last_flush
+        await writer.close_temporary_file()
```

Comment:
> I assume the Parquet writer would always use multiple files

Not in S3: we upload a single Parquet file, so we have to continue flushing the same file.

That being said, I do agree that some of this could be incorporated back to the writer. I couldn't come up with a pythonic way of optionally "resetting" the context manager, but it is something we can definitely give more thought.

Reply:
> Not in S3: we upload a single Parquet file, so we have to continue flushing the same file.

Ah yes, sorry, forgot about that.

In any case, main thing is to fix the bug - we can look at making the code nicer later :)